core/common/src/ContourDisplay.ts

```diff
@@ -0,0 +1,263 @@
+  export interface TerrainProps {
+    /** See [[Terrain.contourDef]]. */
+    contourDef?: ContourProps;
+    /** See [[Terrain.subCategories]]. */
+    subCategories?: Id64String[];
+  }
+
+  /** Contains a description of how contours should appear for a particular set of subcategories. A terrain is an organizational concept which associates a contour appearance with a list of subcategories within an iModel.
+   * @see [[ContourDisplay.Contour]]
+   */
+  export class Terrain {
+    /** A [[ContourDisplay.Contour]] object describing how the contours for this terrain should appear. Defaults to an instantation of [[ContourDisplay.Contour]] using all of its own default properties. */
+    public readonly contourDef: Contour;
+    /** List of subcategory IDs to which this terrain styling will be applied. */
+    public readonly subCategories: Id64String[];
+
+    public equals(other: Terrain | undefined): boolean {
+      if (this === undefined && other === undefined)
+        return true;
+      if (this === undefined || other === undefined)
+        return false;
+      if (!this.contourDef.equals(other.contourDef))
+        return false;
+      if (this.subCategories.length !== other.subCategories.length)
+        return false;
+      for (const subCategory of this.subCategories) {
+        const match = other.subCategories.find((element) => element === subCategory);
+        if (!match)
+          return false;
+      }
+      return true;
+    }
+
+    private constructor(json?: TerrainProps) {
+      if (undefined === json) {
+        this.contourDef = Contour.fromJSON({});
+        this.subCategories = [];
+      } else {
+        this.contourDef = json.contourDef ? Contour.fromJSON(json.contourDef) : Contour.fromJSON({});
+        this.subCategories = json.subCategories ? [...json.subCategories] : [];
+      }
+    }
+
+    public static fromJSON(json?: TerrainProps) {
+      return json ? new Terrain(json) : new Terrain({});
+    }
+
+    public toJSON(): TerrainProps {
+      const props: TerrainProps = {};
+
+      if (!this.contourDef.equals(Contour.defaults))
+        props.contourDef = this.contourDef.toJSON();
+
+      if (this.subCategories.length > 0)
+        props.subCategories = [...this.subCategories];
+
+      return props;
+    }
+  }
+
+  /** JSON representation of the contour display setup of a [[DisplayStyle3d]]. */
+  export interface SettingsProps {
+    /** See [[ContourDisplay.terrains]]. */
+    terrains?: (TerrainProps | undefined)[];
+  }
+
+  /** The contour display setup of a [[DisplayStyle3d]].
+   * Contour display allows a user to apply specific contour line renderings to subcategories within a scene.
+   */
+  export class Settings {
+    /** A list of the terrains which contain their own specific contour display settings. Defaults to an empty array. */
+    public readonly terrains: (Terrain | undefined)[] = [];
```

Comment:
Why does it explicitly permit a sparse array?

Reply:
So that the contour definitions can be used independently (and cleared, by having no subcategories) without them moving around.

Reply:
Was that done to make it easier to implement the UI in display-test-app? It seems harder to use than a `ReadonlyArray<Terrain>`.

Reply:
It was done mainly to make it easier to use (not implement) the UI, and it seemed to me likely the same would be true for Civil.  Currently we don't have a name for a particular contour definition group, just a number, so when you want to change something you select that number and change the definition (color, whatever).  You quickly remember what number goes with what group, so if that was changing whenever you deleted a lower one (or whatever else might rearrange them), that would get very confusing to use.  (To me, identifying a group by remembering a list of subCategoryIds is very unlikely.)

Reply:
Probably should ask Civil for examples of how their UI for this stuff works in their existing app.

Reply:
Based on a side conversation, it was proposed by @pmconne to add a "name" property to each ContourGroup which identifies them. That way they can be kept track of without the need for a sparse array.

Reply:
So will we allow creation of a ContourGroup with no / empty subCategories now?

Reply:
I think there are two separate questions.
1. Is empty subcategory list allowed? If not, you'd have to throw an exception or silently ignore when people try to create one. I think it's better to accept it.
2. What does empty subcategory list signify? Does it mean no subcategories will have this contour style applied, or does it mean this is the default style that applies to subcategories not included in other contour groups? If the latter, and we decide later to add support for a default contour group, we'll need a different way to indicate that. But you would be able to punt for now on adding support for default group.

I vote for whichever is more expedient. Is default group difficult or time-consuming?

Reply:
For reference, [ModelClipGroup](https://github.com/iTwin/itwinjs-core/blob/8ab7ad738c5127c3bdc874ba4019b9e3c54d0171/core/common/src/ModelClipGroup.ts#L28) interprets an empty list of models as indicating the clip should apply to all otherwise-ungrouped models.

Reply:
What is the difference in meaning between "supporting a default group" and "supporting using a ContourGroup with an empty subcategory list as a default style for subcategories not included in other contour groups"?

I guess I am unsure of the definition of the first.  (The latter I think would not be difficult, but we probably also would want a switch to enable/disable, and if there is more than one defined this way, we just use one.)

Reply:
> What is the difference in meaning between "supporting a default group" and "supporting using a ContourGroup with an empty subcategory list as a default style for subcategories not included in other contour groups"?

The latter is one way of implementing the former. I'm unclear what you mean by "a switch to enable/disable" - are you talking about the UI in display-test-app?

Reply:
I was thinking of some flag in the API (and also the UI I guess) to disable having the default definition applied, kind of like the switch to display contours or not in general.

Reply:
You can turn off the default group by removing the default group, just like any other group.

Reply:
Okay, I just pushed an implementation that uses a definition with an empty subcategory list as the default for any features that don't match any of the subCategoryIds. (If there is more than one definition it uses the last, if there are none, then default is not to draw contours.)

Reply:
Code has been pushed that makes this array of contour groupings no longer a sparse array.
Test UI functions properly with this new paradigm.